pkg/reconciler/apis/permissionclaimlabel/permissionclaimlabel_controller.go

```diff
@@ -0,0 +1,222 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package permissionclaimlabel
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	jsonpatch "github.com/evanphx/json-patch"
+	"github.com/kcp-dev/logicalcluster/v2"
+
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clusters"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	apisv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1"
+	kcpclient "github.com/kcp-dev/kcp/pkg/client/clientset/versioned"
+	apisinformers "github.com/kcp-dev/kcp/pkg/client/informers/externalversions/apis/v1alpha1"
+	apislisters "github.com/kcp-dev/kcp/pkg/client/listers/apis/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/informer"
+)
+
+const (
+	controllerName = "kcp-permissionclaimlabel"
+)
+
+// NewController returns a new controller for APIBindings.
+func NewController(
+	kcpClusterClient kcpclient.Interface,
+	dynamicClusterClient dynamic.Interface,
+	dynamicDiscoverySharedInformerFactory *informer.DynamicDiscoverySharedInformerFactory,
+	apiBindingInformer apisinformers.APIBindingInformer,
+) (*controller, error) {
+	queue := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), controllerName)
+
+	c := &controller{
+		queue:                queue,
+		kcpClusterClient:     kcpClusterClient,
+		dynamicClusterClient: dynamicClusterClient,
+		ddsif:                dynamicDiscoverySharedInformerFactory,
+
+		apiBindingsLister: apiBindingInformer.Lister(),
+		listAPIBindings: func(clusterName logicalcluster.Name) ([]*apisv1alpha1.APIBinding, error) {
+			list, err := apiBindingInformer.Lister().List(labels.Everything())
+			if err != nil {
+				return nil, err
+			}
+
+			var ret []*apisv1alpha1.APIBinding
+
+			for i := range list {
+				if logicalcluster.From(list[i]) != clusterName {
+					continue
+				}
+
+				ret = append(ret, list[i])
+			}
+
+			return ret, nil
+		},
+		apiBindingsIndexer: apiBindingInformer.Informer().GetIndexer(),
+	}
+
+	apiBindingInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(obj interface{}) { c.enqueueAPIBinding(obj, "") },
+		UpdateFunc: func(_, obj interface{}) { c.enqueueAPIBinding(obj, "") },
```

Comment:
I am wondering if we can wait on this.

It seems to me that to determine if we need to update, we need to calculate removals and additions and because that is what the reconciler does, I am having a hard time following the logic that should be here. 

I assume in the future, with rejected and awaiting approval, we will need to handle things a little differently but wondering if the set functions can wait for the type updates.